### PR TITLE
development

### PR DESCRIPTION
**Pull request by Kasiama**
First PR

### CODE REVIEW

File: PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+import UIKit
+
+public typealias ConfigureLinkAttribute = ([NSAttributedString.Key: Any], Bool) -> ([NSAttributedString.Key: Any])
+
+class ActiveLabel: UILabel {
+
+    lazy var activeElements = [ElementTuple]()
+
+    fileprivate var _customizing: Bool = true
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift]: @Kasiama naming?

File: PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        paragraphStyle.alignment = textAlignment
+        attributes[NSAttributedString.Key.paragraphStyle] = paragraphStyle
+        mutAttrString.setAttributes(attributes, range: range)
+
+        return mutAttrString
+    }
+
+    fileprivate func parseTextAndExtractActiveElements(_ attrString: NSAttributedString) -> String {
+        let textString = attrString.string
+        let filter: ((String) -> Bool)? = nil
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift]: @Kasiama do this variable needed?

File: PhotoApp/MainViewController/FullImageViewController.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class FullImageViewController: UIViewController {
+
+    @IBOutlet weak var descriptionLabel: ActiveLabel!
+
+    @IBOutlet  var imageView: CachedImageView!
+    var image: UIImage?
+    var photoDescription = ""
+    var hastags: [String]?
+    var date: String?
+    var searchbar = UISearchBar.init()
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/FullImageViewController.swift]: @Kasiama maybe it's better to initialize searchbar in "init" of the class? :) Your thoughts

File: PhotoApp/MainViewController/FullImageViewController.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+//  Copyright © 2019 Ivan Zhurauski. All rights reserved.
+//
+
+import UIKit
+
+class FullImageViewController: UIViewController {
+
+    @IBOutlet weak var descriptionLabel: ActiveLabel!
+
+    @IBOutlet  var imageView: CachedImageView!
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/FullImageViewController.swift]: @Kasiama does this variable need to be strong?

File: PhotoApp/MainViewController/FullImageViewController.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        photoDescription = description ?? ""
+            load()
+    }
+
+    func load() {
+        self.view.backgroundColor = UIColor.gray
+        descriptionLabel.customize { label in
+            label.text = " " + photoDescription
+            label.numberOfLines = 0
+            label.textColor = UIColor(red: 255, green: 255, blue: 255, alpha: 1)
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/FullImageViewController.swift]: @Kasiama UIColor.white?

File: PhotoApp/MainViewController/FullImageViewController.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            label.hashtagColor = UIColor(red: 85.0/255, green: 172.0/255, blue: 238.0/255, alpha: 1)
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/FullImageViewController.swift:36]: @Kasiama just a note for the future: it's good to create some color constants or use Colors.xcasset with color manager. If interested — ask about it :)

File: PhotoApp/MainViewController/FullImageViewController.swift [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        self.imageView.layer.addSublayer(gradient)
+
+        let gradientе = CAGradientLayer()
+        let boundss = CGRect.init(x: 0, y: UIScreen.main.bounds.height - UIScreen.main.bounds.height/5, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height/5)
+        gradientе.frame = boundss
+        gradientе.colors = [UIColor.black.cgColor, UIColor.clear.cgColor]
+        gradientе.startPoint = CGPoint(x: 0, y: 1)
+        gradientе.endPoint = CGPoint(x: 0, y: 0)
+        self.imageView.layer.addSublayer(gradientе)
+        }
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/FullImageViewController.swift:57]: @Kasiama spacing between methods? :) No worries, just a note

File: PhotoApp/MainViewController/MainViewController.swift [hunk 8]
```diff
@@ -131,269 +154,246 @@ class MainViewController: UIViewController,UIGestureRecognizerDelegate,UIImagePi
...
     func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
         self.dismiss(animated: true, completion: nil)
     }
-    
-    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
-        self.dismiss(animated: true, completion: nil)
+
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
+
+        var currentdate: Date!
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/MainViewController.swift]: @Kasiama could you explain, please var and "!"?

File: PhotoApp/MainViewController/PopupViewController.swift [hunk 9]
```diff
@@ -10,208 +10,222 @@ import UIKit
 import MapKit
 import Firebase
 
-
-
 extension String {
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/PopupViewController.swift]: @Kasiama important: **always move such generic extensions to separated files**

File: PhotoApp/MainViewController/PopupViewController.swift [hunk 10]
```diff
@@ -10,208 +10,222 @@ import UIKit
...
-            photoModel = Photomodel.init(id: key, latitude: self.annotation.coordinate.latitude, longitude:self.annotation.coordinate.longitude, category: cat, date: "", hashtags: hashtags, description: self.descriptionTextView.text, image: UIImage.init())
-            let photomodelsend = [ "photoID":  key,
-                                   "latitude": photoModel!.latitude,
-                                   "longitude": photoModel!.longitude,
-                                   "date": photoModel!.date,
-                                   "hashtags": photoModel!.hashtags,
-                                   "description": photoModel!.description,
-                                   "categoryID": cat.id] as [String : Any]
-                var childUpdates = ["/\(String(describing: Auth.auth().currentUser!.uid))/photomodels/\(key)": photomodelsend]
+            } else {
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/PopupViewController.swift]: @Kasiama such if can be placed at one line :) just a note
example If 1 {
               } else if let .... {
               }

**Review by DerHlopec — CHANGES_REQUESTED** [review]
Please, recheck comments :)

File: PhotoApp/MainViewController/Other/OtherViewController.swift [hunk 11]
```diff
@@ -32,15 +56,68 @@ class OtherViewController: UIViewController {
...
 
-    /*
-    // MARK: - Navigation
+            }
+        }
+    }
+
+    func openGallery() {
+        let status = PHPhotoLibrary.authorizationStatus()
+        if status == .denied {} else {
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/Other/OtherViewController.swift]: @Kasiama if status == .authorized? :D But what about another cases: .restricted, .notDetermined? In such cases we should provide an option to request this authority status :)

**Review by DerHlopec — CHANGES_REQUESTED** [review]

File: PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift (hunk 1 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift]: fixed

File: PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift (hunk 2 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/ActiveLabel/ActiveLabel.swift]: fixed

File: PhotoApp/MainViewController/FullImageViewController.swift (hunk 3 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/FullImageViewController.swift]: we don't need this property)

File: PhotoApp/MainViewController/FullImageViewController.swift (hunk 4 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/FullImageViewController.swift]: no, fixed

File: PhotoApp/MainViewController/FullImageViewController.swift (hunk 5 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/FullImageViewController.swift]: okay) Fixed

File: PhotoApp/MainViewController/MainViewController.swift (hunk 8 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/MainViewController.swift]: I just tested something. Fixed

File: PhotoApp/MainViewController/PopupViewController.swift (hunk 9 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/PopupViewController.swift]: fixed

File: PhotoApp/MainViewController/PopupViewController.swift (hunk 10 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/PopupViewController.swift]: fixed, will know

File: PhotoApp/MainViewController/Other/OtherViewController.swift (hunk 11 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/Other/OtherViewController.swift]: fixed

File: PhotoApp/MainViewController/Other/AboutViewController.swift [hunk 12]
```diff
@@ -10,10 +10,21 @@ import UIKit
...
-
+        
+        self.imageView.image = UIImage.init(named: "me")
+        self.imageView.contentMode = .scaleAspectFill
         // Do any additional setup after loading the view.
     }
 
+    @IBAction func vkbuttonTaped(_ sender: Any) {
+        UIApplication.shared.open(NSURL(string:"https://vk.com/fukakasiama")! as URL)
+    }
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/Other/AboutViewController.swift:29]: @Kasiama I know, that you are totally sure, that this is URL, but could you please release it with optional with our force unwrapping?

File: PhotoApp/MainViewController/Other/OtherViewController.swift [hunk 13]
```diff
@@ -85,7 +94,14 @@ class OtherViewController: UIViewController, UIImagePickerControllerDelegate, UI
 
     func openGallery() {
         let status = PHPhotoLibrary.authorizationStatus()
-        if status == .denied {} else {
+        if status == .denied {
```
Comment (DerHlopec) [inline PhotoApp/MainViewController/Other/OtherViewController.swift:109]: @Kasiama copy-paste code? :) Just a note

**Review by DerHlopec — CHANGES_REQUESTED** [review]
Last fixes

File: PhotoApp/MainViewController/Other/AboutViewController.swift (hunk 12 above)
Comment (Kasiama) [reply PhotoApp/MainViewController/Other/AboutViewController.swift:29]: okey, no problem)

**Review by DerHlopec — APPROVED** [review]

**Pull request merged**